SignalServiceKit/src/Loki/Extensions/SessionCipher+Loki.m

```diff
@@ -0,0 +1,181 @@
+    if (!state) {
+        [self throws_verifyFriendRequestAcceptPreKeyForMessage:whisperMessage protocolContext:protocolContext];
+    }
+   
+    // Loki: Handle any session resets
+    [self handleSessionReset:whisperMessage previousState:state protocolContext:protocolContext];
+    
+    return plainText;
+}
+
+/// Get the current session state
+- (SessionState *_Nullable)getCurrentState:(nullable id)protocolContext {
+    SessionRecord *record = [self.sessionStore loadSession:self.recipientId deviceId:self.deviceId protocolContext:protocolContext];
+    SessionState *state = record.sessionState;
+    
+    // Check if session is initialized
+    if (!state.hasSenderChain) {
+        return nil;
+    }
+    
+    return state;
+}
+
+/// Handle any loki session reset stuff
+- (void)handleSessionReset:(id<CipherMessage>)whisperMessage
+             previousState:(SessionState *_Nullable)previousState
+           protocolContext:(nullable id)protocolContext
+{
+    // Don't bother doing anything if we didn't have a session before
+    if (!previousState) {
+        return;
+    }
+    
+    OWSAssertDebug([protocolContext isKindOfClass:[YapDatabaseReadWriteTransaction class]]);
+    YapDatabaseReadWriteTransaction *transaction = protocolContext;
+    
+    // Get the thread
+    TSContactThread *thread = [TSContactThread getThreadWithContactId:self.recipientId transaction:transaction];
+    if (!thread) {
+        return;
+    }
+    
+    // Bail early if no session reset is in progress
+    if (thread.sessionResetState == TSContactThreadSessionResetStateNone) {
+        return;
+    }
+    
+    BOOL sessionResetReceived = thread.sessionResetState == TSContactThreadSessionResetStateRequestReceived;
+    SessionState *_Nullable currentState = [self getCurrentState:protocolContext];
+    
+    // Check if our previous state and our current state differ
+    if (!currentState || ![currentState.aliceBaseKey isEqualToData:previousState.aliceBaseKey]) {
+        
+        if (sessionResetReceived) {
+            // The other user used an old session to contact us.
+            // Wait for them to use a new one
+            [self restoreSession:previousState protocolContext:protocolContext];
+        } else {
+            // Our session reset went through successfully
+            // We had initiated a session reset and got a different session back from the user
+            [self deleteAllSessionsExcept:currentState protocolContext:protocolContext];
+            [self notifySessionAdopted];
+        }
+        
+    } else if (sessionResetReceived) {
+        // Our session reset went through successfully
+        // We got a message with the same session from the other user
+        [self deleteAllSessionsExcept:previousState protocolContext:protocolContext];
+        [self notifySessionAdopted];
+    }
+}
+
+/// Send a notification about a new session being adopted
+- (void)notifySessionAdopted
+{
+    [[NSNotificationCenter defaultCenter]
```

Comment:
I'm thinking we may need to flush all transactions before posting this notification to avoid threading issues